packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,218 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  Directive,
+  PartInfo,
+  $private,
+  NodePart,
+  AttributePart,
+  Part,
+} from './lit-html.js';
+export {directive} from './lit-html.js';
+
+const {_TemplateInstance: TemplateInstance} = $private;
+type TemplateInstance = InstanceType<typeof TemplateInstance>;
+
+type DisconnectableChild =
+  | NodePart
+  | AttributePart
+  | TemplateInstance
+  | DisconnectableDirective;
+
+// Contains the sparse tree of Parts/TemplateInstances needing disconnect
+const disconnectableChildrenForParent: WeakMap<
+  DisconnectableChild,
+  Set<DisconnectableChild>
+> = new WeakMap();
+
+/**
+ * Recursively walks down the tree of Parts/TemplateInstances to set the
+ * connected state of directives and run `disconnectedCallback`/
+ * `reconnectedCallback`s.
+ */
+const setConnected = (
+  child: DisconnectableChild,
+  isConnected: boolean,
+  shouldRemoveFromParent = false
+) => {
+  const children = disconnectableChildrenForParent.get(child);
+  if (children !== undefined) {
+    for (const child of children) {
+      if (child instanceof DisconnectableDirective) {
```

Comment:
The `instanceof` checks worry me here. If we somehow get multiple instances of this module, then this function could encounter `DisconnectableDirective` from the other instance and the check will fail. Same for the `instanceof NodePart`, etc checks.

Reply:
Yup, there's a few instances of multiple-version safety that I still need to work out; we can do the branding thing here, but  I wanted to decide on overall strategy here first before fixing them all; if we want to put all this code in the core, it's moot.